### PR TITLE
Student's dashboard rank list

### PR DESCRIPTION
**Pull request by realARCHaim**
https://trello.com/c/JgyqrUBn

### CODE REVIEW

File: askup/utils/general.py [hunk 1]
```diff
@@ -92,6 +116,69 @@ def get_user_place_in_rank_list(user_id):
...
+
+
+def compose_user_profile_rank_list_row_data(rows, user_id_to_check=None):
+    """
+    Compose user profile rank list row datas from the query result rows.
+    """
+    items = []
+    user_is_present = False
+
+    for place, id, username, first_name, last_name, questions, thumbs_up in rows:
```
Comment (cmltaWt0) [inline askup/utils/general.py]: Please do not redefine built-in `id`

File: askup/utils/general.py [hunk 2]
```diff
@@ -92,6 +116,69 @@ def get_user_place_in_rank_list(user_id):
...
+    """
+    Return user's full name representation for the views.
+
+    Needed because of unrequired first and last names.
+    """
+    name = (first_name or last_name) and ' '.join((first_name, last_name))
+    return '{} ({})'.format(name, username) if name else username
+
+
+def get_user_profile_rank_list_elements(viewer_user_id, expression='limit 10', args=[]):
```
Comment (cmltaWt0) [inline askup/utils/general.py]: You mustn't define default value for arg as `[]`. It leads to bad situations.

File: askup/utils/general.py [hunk 3]
```diff
@@ -369,3 +456,12 @@ def get_student_last_week_incorrect_answers_count(user_id):
...
 
     return 0
+
+
+def get_user_organizations_string(user):
+    """
+    Return comma separated user organizations string.
+    """
+    return ', '.join(
+        (str(org) for org in askup.models.Organization.objects.filter(users__in=[user]))
```
Comment (cmltaWt0) [inline askup/utils/general.py:478]: Why do you use `__in` instead of simple equal?

File: askup/views.py [hunk 4]
```diff
@@ -257,6 +258,39 @@ def user_profile_view(request, user_id):
...
+@login_required
+def user_profile_rank_list_view(request, user_id):
+    """Provide the user profile rank list view."""
+    profile_user = get_object_or_404(User, pk=user_id)
+    rank_list = get_user_profile_rank_list(profile_user.id, request.user.id)
+    return render(
+        request,
+        'askup/user_profile.html',
+        {
+            'profile_user': profile_user,
```
Comment (cmltaWt0) [inline askup/views.py:276]: Please use `user_id ` instead of `profile_user.id` each time.

File: askup/templates/askup/user_profile.html [hunk 5]
```diff
@@ -2,15 +2,18 @@
...
+        <h3>{{ user_organizations|default:"No organizations assigned to the user" }}</h3>
+    </div>
     <div class="row">
         <div class="col-sm-3">
-            <div class="box center dash-profile">
+            <div class="box center dash-profile rounded-10">
                 <img alt="Person" src="/static/assets/default-avatar.png">
                 <br><br>
-                <h3 class="center">{{ first_name }} {{ last_name }}</h3>
+                <h3 class="center">{{ profile_user.first_name }} {{ profile_user.last_name }}</h3>
```
Comment (cmltaWt0) [inline askup/templates/askup/user_profile.html]: What is a user has no First name AND Last name ?

File: askup/utils/general.py [hunk 6]
```diff
@@ -92,6 +116,69 @@ def get_user_place_in_rank_list(user_id):
...
+    Aquire and return a list of rows for the user profile rank list.
+    """
+    first_items = get_user_profile_rank_list_elements(viewer_user_id, args=(rank_user_id,))
+    result_row_datas, user_is_present = compose_user_profile_rank_list_row_data(first_items, rank_user_id)
+
+    if user_is_present:
+        return result_row_datas
+
+    result_row_datas += [(0, None, None, None, None)]  # adding an ellipsis row
+    user_item = get_user_profile_rank_list_elements(rank_user_id, 'where ranked.id = %s', (rank_user_id,))
```
Comment (cmltaWt0) [inline askup/utils/general.py:130]: As for me it is completely wrong solution for this project and for actually Django. But it is an architectural suggestion only.

**Review by cmltaWt0 — COMMENTED** [review]
My prevailing concern is about raw SQL. So further decisions should on the project tech lead.
All my comments are simply suggestions.
Thanks.

File: askup/utils/general.py (hunk 1 above)
Comment (realARCHaim) [reply askup/utils/general.py]: thanks. did it.

File: askup/utils/general.py (hunk 3 above)
Comment (realARCHaim) [reply askup/utils/general.py:478]: thank you. fixed. didn't know that you can compare a single object here to the set of multiple.

File: askup/views.py (hunk 4 above)
Comment (realARCHaim) [reply askup/views.py:276]: done

File: askup/templates/askup/user_profile.html (hunk 5 above)
Comment (realARCHaim) [reply askup/templates/askup/user_profile.html]: fixed this moment by using the function that returns the full user name representation
`[<first name> <last name> (]<username>[)]`

File: askup/utils/general.py (hunk 6 above)
Comment (realARCHaim) [reply askup/utils/general.py:130]: Do you mean raw sql?

File: askup/utils/general.py (hunk 3 above)
Comment (cmltaWt0) [reply askup/utils/general.py:478]: @KhaimovMR What did you fix? I see this comment and not sure I am right.

File: askup/utils/general.py (hunk 6 above)
Comment (cmltaWt0) [reply askup/utils/general.py:130]: @KhaimovMR You pass part or raw SQL query here.

File: askup/utils/general.py (hunk 3 above)
Comment (realARCHaim) [reply askup/utils/general.py:478]: Yeah. I've tried to compare a users set to a single user object:
`filter(users=user)`
and it worked

File: askup/utils/general.py (hunk 6 above)
Comment (realARCHaim) [reply askup/utils/general.py:130]: Oh yes. It's made in the DRY concerns. I can make two similar functions but with different where expressions in this case.

File: askup/utils/general.py (hunk 2 above)
Comment (realARCHaim) [reply askup/utils/general.py]: done

File: askup/utils/general.py (hunk 6 above)
Comment (idegtiarov) [reply askup/utils/general.py:130]: We will leave all as it is now, and return to this part in refactoring in one of the next iterations.

**Review by idegtiarov — APPROVED** [review]

**Pull request merged**